### ISSUE TITLE
gzip compression will post empty files to S3 bucket

### ISSUE THREAD

**Issue by lremurphy** [logstash-plugins/logstash-output-s3#95]
When using:
encoding => "gzip" 

in the output filter for logstash, it will post empty 20 byte files to S3 in concert with the time_file parameter set at 10 minute intervals instead of skipping the time interval if the file is empty.

**Comment by PedroMDuarte** [comment]
I wonder if this could be dealt with by changing the definition of `size` on the `IOWrappedGzip`:

https://github.com/logstash-plugins/logstash-output-s3/blob/dbc474bd54b64ea985ea59f70777d6d13c5be211/lib/logstash/outputs/s3/temporary_file_factory.rb#L110

perhaps one could use `@gzip_writer.pos()` instead of `flush` followed by `to_io.size`. 

See the documentation for `pos()` here: https://ruby-doc.org/stdlib-1.9.3/libdoc/zlib/rdoc/Zlib/GzipWriter.html

This is the first time I look closely into ruby source code, so don't want to venture on writing tests and doing a pull request for this, just posting this here to check if my understanding is correct.

**Comment by ph** [comment]
@PedroMDuarte It is still an issue? which version of the plugin/logstash you are using?

**Comment by PedroMDuarte** [comment]
Hey, thanks for the quick reply! 

Thanks for the quick reply! 

```
$ find . -name '*s3*' | grep logstash-output
./vendor/bundle/jruby/1.9/specifications/logstash-output-s3-4.0.5.gemspec
./vendor/bundle/jruby/1.9/gems/logstash-output-s3-4.0.5
./vendor/bundle/jruby/1.9/gems/logstash-output-s3-4.0.5/lib/logstash/outputs/s3
./vendor/bundle/jruby/1.9/gems/logstash-output-s3-4.0.5/lib/logstash/outputs/s3.rb
./vendor/bundle/jruby/1.9/gems/logstash-output-s3-4.0.5/logstash-output-s3.gemspec
```
Yes it is still an issue, it saves empty gzip files that are 26 Bytes in size.

**Comment by PedroMDuarte** [comment]
for example:

```
$ aws s3 ls s3://***/logstash-throughput/20170215_07/ --human-readable
2017-02-15 00:55:45   26 Bytes ls.s3.42e81ae0-84c3-4824-a98c-ad8c55c841ac.2017-02-15T08.35.part2.txt.gz
2017-02-15 00:15:45  593.0 KiB ls.s3.783dfe96-52f7-4a1e-a008-6abd221def22.2017-02-15T07.55.part0.txt.gz
2017-02-15 00:35:45   26 Bytes ls.s3.7e85cac4-0350-4a0a-a450-f3a7be565307.2017-02-15T08.15.part1.txt.gz
2017-02-15 01:15:46   26 Bytes ls.s3.d7fb2275-6164-400e-9d96-ee5f3ddc5baf.2017-02-15T08.55.part3.txt.gz
```

Here is the configuration that I am using:

```
s3 {
    access_key_id => "{{ aws_key }}"
    secret_access_key => "{{ aws_secret }}"
    bucket => "{{ aws_bucket }}"
    canned_acl => "private"
    codec => "json_lines"
    enable_metric => true
    encoding => "gzip"
    id => "rabbitmq_logstash_s3_output"
    prefix => "logstash-throughput/%{+YYYYMMdd_HH}"
    region => "us-west-2"
    size_file => 15242880
    time_file => 20
    rotation_strategy => "size_and_time"
}
```

**Comment by ph** [comment]
definitively something is wrong with the #Size implementation on a temporary gzip file, the time rotation uses the right call.

https://github.com/ph/logstash-output-s3/blob/e5ee0bd1f4c412a68eb51e328cbdeb2479e56682/lib/logstash/outputs/s3/time_rotation_policy.rb#L17

**Comment by ph** [comment]
@PedroMDuarte I think your idea of using `#pos` will work in that case, it might be even enough to just use `#pos` and not do any flush on the io, the uploader will still force a flush before uploading.

+1 to make a PR and I will review it.

**Comment by PedroMDuarte** [comment]
will do,  I'll see if I can manage to add a test for that.  first time using Ruby ...

**Comment by PedroMDuarte** [comment]
PR is ready here https://github.com/logstash-plugins/logstash-output-s3/pull/128

**Comment by ph** [comment]
fixed by #128
